### ISSUE TITLE
Error while building in unreal engine 4.27

### ISSUE THREAD

**Issue by awsproserve** [AppLovin/AppLovin-MAX-Unreal#13]
### MAX Plugin Version

1.1.2

### Unreal Engine Version

4.27

### Device/Platform Info

IOS

### Current Behavior

I install the plugin to unreal engine and while trying to build the application I get the following error below. Been trying for all this week and still no avail, any help in the right direction would be greatly appreciated. Thanks - Sarah

![image](https://user-images.githubusercontent.com/116039531/202844995-2e2a09e1-1ac8-42c0-9dfd-a2c64d3b58fb.png)

### Expected Behavior

The game/application should build successfully without error, and package successfully

### How to Reproduce

1. First lets down load the plugin. I navigate to the GIT repo, and download the plugin folder
![image](https://user-images.githubusercontent.com/116039531/202845116-096d1764-3925-4bdc-bae2-9ccbbdd512ef.png)

2. Go to your downloads and find the folder we just downloaded and copy the plugin from the plugin folder
![image](https://user-images.githubusercontent.com/116039531/202845201-8a517619-259e-4547-aee7-b5cc8a19fdcd.png)

3. Put the applovin max plugin in your projects plugin folder
![image](https://user-images.githubusercontent.com/116039531/202845245-aa7e0bc4-46d1-4ab1-9a36-4c2e172694f0.png)

4. Find your [yourgame/appname].target.cs file within your project folder, open with text editor and add the following below

 // Add compiler argument to support building on macOS with Xcode 13.3+
    bOverrideBuildEnvironment = true;
    AdditionalCompilerArguments = "-Wno-unused-but-set-variable";

![image](https://user-images.githubusercontent.com/116039531/202845705-a78038bc-a7cb-4154-aefd-9dafe52027eb.png)

5. Add AppLovinMAX as a dependency to the PublicDependencyModuleNames array inside of your app’s <PROJECTNAME>.Build.cs file:

Add AppLovinMAX as a dependency to the PublicDependencyModuleNames array inside of your app’s <PROJECTNAME>.Build.cs file:
for example:
PublicDependencyModuleNames.AddRange(new string[] { "Core", "CoreUObject", "Engine", "InputCore", "AppLovinMAX" });

Screenshot of me adding it to my app, our public dependency module name line is formatted by default different than yours, so I used it the way my project has it in in default with adding the applovin, input core, engine, and coreuobject. and I also tired it with your exact formate. see below
 |
V
![image](https://user-images.githubusercontent.com/116039531/202846118-664ca8f9-97e5-4810-a897-3600079eeb19.png)

(I just noticed that after I save this file and re open it, my changes save to the run.build.cs file, however when I try to package the app, and it fails, I go back to the run.build.cs file and it reverts back to default, is that expected? see below for the default that it reverts back to 

![image](https://user-images.githubusercontent.com/116039531/202846170-7a760f13-6c21-4f8d-9b58-7afac44e0e54.png)

now I try it exactly the way you have it in your instructions, and tried building package and still not working. it still also reverts back to default when package fails. 

![image](https://user-images.githubusercontent.com/116039531/202847099-3da255db-7312-4f67-8198-03f49d9e0305.png)

6. Update your app’s Info.plist in Unreal’s Project Settings with network-specific identifiers. See the [SKAdNetwork documentation](https://dash.applovin.com/documentation/mediation/unreal/getting-started/skadnetwork-info) for instructions.

![image](https://user-images.githubusercontent.com/116039531/202846256-75a015d5-999b-40f6-996d-7fe5125c62bf.png)

7. Open your project, to go file > package > IOS

<img width="981" alt="image" src="https://user-images.githubusercontent.com/116039531/202846298-a6dac67f-53a7-4e5a-b358-5c47176ead0f.png">

8. After few seconds, I get the following error 

to suppress this warning.
UATHelper: Packaging (iOS):   Couldn't find PLCrashReporter in folder 'lib-Xcode-13.4.1', using default 'lib-Xcode-12.4'
UATHelper: Packaging (iOS):   Compiling against OS Version 12.0 [minimum allowed at runtime]
UATHelper: Packaging (iOS):   Parsing headers for run
UATHelper: Packaging (iOS):     Running UnrealHeaderTool "/Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/run.uproject" "/Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Intermediate/Build/IOS/run/Shipping/run.uhtmanifest" -LogCmds="loginit warning, logexit warning, logdatabase error" -Unattended -WarningsAsErrors -absl
og="/Users/caataiwi/Library/Logs/Unreal Engine/LocalBuildLogs/UHT-run-IOS-Shipping_2.txt" -installed
UATHelper: Packaging (iOS):   2022-11-19 02:18:44.600 UnrealHeaderTool[80304:7868319] [UE4] Looking for binary: ../../../Engine/Programs/UnrealHeaderTool/Config/BinaryConfig.ini
UATHelper: Packaging (iOS):   LogInit: Display: Loading text-based GConfig....
UATHelper: Packaging (iOS):   Reflection code generated for run in 4.5829604 seconds
UATHelper: Packaging (iOS):   Compiling with iPhoneOS SDK 15.5
UATHelper: Packaging (iOS):   Zip file: /Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Binaries/IOS/run-IOS-Shipping.dSYM.zip
UATHelper: Packaging (iOS):   Adding PostBuildSync action
UATHelper: Packaging (iOS):   IOSPlugin: Plugins/Runtime/AR/AppleAR/AppleARKit/Source/AppleARKit/AppleARKit_IOS_UPL.xml
UATHelper: Packaging (iOS):   IOSPlugin: Plugins/Marketplace/EasyAdsPro/Source/EasyAds/EasyAds_IOS_UPL.xml
UATHelper: Packaging (iOS):   IOSPlugin: Plugins/Marketplace/PsIronSource/Source/PsIronSource/IronSource_APL_IOS.xml
UATHelper: Packaging (iOS):   IOSPlugin: ../../../caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Plugins/AppLovinMAX/Source/AppLovinMAX/AppLovinMAX_UPL_IOS.xml
UATHelper: Packaging (iOS):   Writing manifest to /Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Intermediate/Build/Manifest.xml
UATHelper: Packaging (iOS):   ERROR: Unable to merge actions producing AppLovinSDK.embeddedframework.extracted: prerequisites are different
PackagingResults: Error: Unable to merge actions producing AppLovinSDK.embeddedframework.extracted: prerequisites are different
UATHelper: Packaging (iOS):   ERROR: Action graph is invalid; unable to continue. See log for additional details.
PackagingResults: Error: Action graph is invalid; unable to continue. See log for additional details.
UATHelper: Packaging (iOS): Took 9.091533s to run mono, ExitCode=6
UATHelper: Packaging (iOS): UnrealBuildTool failed. See log for more details. (/Users/caataiwi/Library/Logs/Unreal Engine/LocalBuildLogs/UBT-run-IOS-Shipping_2.txt)
UATHelper: Packaging (iOS): AutomationTool exiting with ExitCode=6 (6)
UATHelper: Packaging (iOS): RunUAT ERROR: AutomationTool was unable to run successfully.
PackagingResults: Error: AutomationTool was unable to run successfully.
PackagingResults: Error: Unknown Error
LogSlate: Window 'Output Log' being destroyed
LogSlate: Window 'Output Log' being destroyed

![image](https://user-images.githubusercontent.com/116039531/202846335-0da4bdfd-6e84-4b34-8880-f4e184f5ffcd.png)

Additional note: When I remove the plugin and revert all my changes I made, I can successfully package the IOS  and upload it to apples test flight. 

If anyone can please help us with what we are doing wrong here. I want to note we have't yet implemented any blueprints or code specific to the plugin, we wanted to make sure that we can package the game after we installed the plugin to make sure nothing broke. Thanks in advance for any help you can provide. :)

**Comment by ritamsarmah** [comment]
Hi @awsproserve, thank you for the detailed screenshots. I think this error may be due to the incorrect formatting in the `Info.plist`, since the added `SKAdNetworkIdentifier` items should be inside the `array` tag of `SKAdNetworkItems`, for example:

```
<key>SKAdNetworkItems</key>
	<array>
		<dict>
			<key>SKAdNetworkIdentifier</key>
			<string>...</string>
		</dict>
		<dict>
			<key>SKAdNetworkIdentifier</key>
			<string>...</string>
		</dict>
		<dict>
			<key>SKAdNetworkIdentifier</key>
			<string>...</string>
		</dict>
        </array>
```

**Comment by ritamsarmah** [comment]
You can also add the full `SKAdNetworkItems` into `AppLovinMAX_UPL_IOS.xml` if it's easier to format there

**Comment by awsproserve** [comment]
Hello @ritamsarmah you where right, when I tried to open up the info.plist file in Xcode, without opening in text, it threw me an error. I have used your suggested format and then it opened up successfully in Xcode. see screenshot below. I tried to build in iOS though and I got still the same error,  

ERROR:

UATHelper: Packaging (iOS):   AppLovin IOS Plugin found
UATHelper: Packaging (iOS):   Couldn't find PLCrashReporter in folder 'lib-Xcode-13.4.1', using default 'lib-Xcode-12.4'
UATHelper: Packaging (iOS):   Compiling against OS Version 12.0 [minimum allowed at runtime]
UATHelper: Packaging (iOS):   Parsing headers for run
UATHelper: Packaging (iOS):     Running UnrealHeaderTool "/Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/run.uproject" "/Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Intermediate/Build/IOS/run/Shipping/run.uhtmanifest" -LogCmds="loginit warning, logexit warning, logdatabase error" -Unattended -WarningsAsErrors -absl
og="/Users/caataiwi/Library/Logs/Unreal Engine/LocalBuildLogs/UHT-run-IOS-Shipping_2.txt" -installed
UATHelper: Packaging (iOS):   2022-11-21 12:52:49.242 UnrealHeaderTool[83952:8144251] [UE4] Looking for binary: ../../../Engine/Programs/UnrealHeaderTool/Config/BinaryConfig.ini
UATHelper: Packaging (iOS):   LogInit: Display: Loading text-based GConfig....
UATHelper: Packaging (iOS):   Reflection code generated for run in 6.2245425 seconds
UATHelper: Packaging (iOS):   Compiling with iPhoneOS SDK 15.5
UATHelper: Packaging (iOS):   Zip file: /Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Binaries/IOS/run-IOS-Shipping.dSYM.zip
UATHelper: Packaging (iOS):   Adding PostBuildSync action
UATHelper: Packaging (iOS):   IOSPlugin: Plugins/Runtime/AR/AppleAR/AppleARKit/Source/AppleARKit/AppleARKit_IOS_UPL.xml
UATHelper: Packaging (iOS):   IOSPlugin: Plugins/Marketplace/EasyAdsPro/Source/EasyAds/EasyAds_IOS_UPL.xml
UATHelper: Packaging (iOS):   IOSPlugin: Plugins/Marketplace/PsIronSource/Source/PsIronSource/IronSource_APL_IOS.xml
UATHelper: Packaging (iOS):   IOSPlugin: ../../../caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Plugins/AppLovinMAX/Source/AppLovinMAX/AppLovinMAX_UPL_IOS.xml
UATHelper: Packaging (iOS):   Writing manifest to /Users/caataiwi/Documents/Unreal Projects/KZ_4.27_Avatar_nov_17_max/Intermediate/Build/Manifest.xml
UATHelper: Packaging (iOS):   ERROR: Unable to merge actions producing AppLovinSDK.embeddedframework.extracted: prerequisites are different
PackagingResults: Error: Unable to merge actions producing AppLovinSDK.embeddedframework.extracted: prerequisites are different
UATHelper: Packaging (iOS):   ERROR: Action graph is invalid; unable to continue. See log for additional details.
PackagingResults: Error: Action graph is invalid; unable to continue. See log for additional details.
UATHelper: Packaging (iOS): Took 12.747466s to run mono, ExitCode=6
UATHelper: Packaging (iOS): UnrealBuildTool failed. See log for more details. (/Users/caataiwi/Library/Logs/Unreal Engine/LocalBuildLogs/UBT-run-IOS-Shipping_2.txt)
UATHelper: Packaging (iOS): AutomationTool exiting with ExitCode=6 (6)
UATHelper: Packaging (iOS): RunUAT ERROR: AutomationTool was unable to run successfully.
PackagingResults: Error: AutomationTool was unable to run successfully.
PackagingResults: Error: Unknown Error

![image](https://user-images.githubusercontent.com/116039531/203156681-3dec0f9b-e626-4806-b90d-e16fc56f7d08.png)

**Comment by awsproserve** [comment]
screenshot of the error 

![image](https://user-images.githubusercontent.com/116039531/203156970-fe2a856b-221a-43f8-886e-9ff6310067b4.png)

**Comment by awsproserve** [comment]
Any other thoughts as to why im getting this error?

**Comment by awsproserve** [comment]
I just disabled the applovin plugin to make sure its not an error with my application, and the iOS build was successful. let me disable the other plugins and remove thier code and see, will get back shortly

**Comment by awsproserve** [comment]
Hi @ritamsarmah I have disabled the other ad plugins, and it works now, thank you so much for your recommendations I was able to build IOS app successfully.  Question, im looking at the blue prints from your sample project, im looking at all the blueprints, but im not seeing anywhere where I can ad a actual numeric banner id. Here are the blueprints im looking at

![image](https://user-images.githubusercontent.com/116039531/203191522-0700ce52-57b2-41de-b3d9-d2add1136fcf.png)


Im looking at the blueprints and I see where you can load, and show as per screenshot below, but where do I specify the numeric id at?


![image](https://user-images.githubusercontent.com/116039531/203191782-bb4173c8-5f32-49c7-898b-dc644d1fcda2.png)

**Comment by awsproserve** [comment]
Also where is the banner ad going to be actually displayed? I dont see visually where it will show up. is this the placeholder for the banner ad? (as seen in the screenshot below?)

![image](https://user-images.githubusercontent.com/116039531/203193631-4b86f2a9-4a04-44d0-aee1-3f8b1b8efdd7.png)

**Comment by awsproserve** [comment]
in summary, 
1/ where do we specify the banner ad ids
2/ how can we visually place the banner ad on the screen? do we have control where the banner ad will show up, or will it show only at the top or the bottom of the screen, and if so how do we tell the banner ad where it should show itself.

**Comment by ritamsarmah** [comment]
The demo app currently only uses C++, so the actual functions being called are in `BannerWidget.cpp`. The blueprint functions go by the same name as the C++ ones, so you'd be using `Create Banner`, which allows you to specify the ad unit ID and position.

You may like to check out our docs on how to work with banners, including the positioning options available: https://dash.applovin.com/documentation/mediation/unreal/ad-formats/banners

**Comment by ritamsarmah** [comment]
Would you be able to provide any crash logs? Could you check that you're initializing the SDK before calling any of the banner methods?

**Comment by awsproserve** [comment]
Hi @ritamsarmah you where correct, we where not initializing the SKD before calling the banner ad, That is now resolved. We ran into another issue. is there a way we can make a action happen AFTER the interstitial ad is called, whats happening now is the player dies, and after they die they can choose to be revived but they have to watch an interstitial ad when they hit the retry button. the problem is when the interstitial ad is displaying the game continues on to restart while the interstitial ad is going on. Using other plugins they exposed in blueprint a way where we can wait until the interstitial ad is finished/ closed, and then have the an action happen immediatlly after, ie the player comes back to life. Whats happening now is while the interstitial ad is going on, the player comes back and dies again before the interstitial ad I even finished.


here is how our current setup for the interstitial blueprint, pls let us know if we can add anything to implement this solution, many thanks and happy thanksgiving :)
![image](https://user-images.githubusercontent.com/116039531/203896966-d57e06df-428f-4241-be62-684472bfbd44.png)

**Comment by awsproserve** [comment]
here is an example of another plugin we used where it gave us the option to do a certain action, after the user closed the ad 
![image](https://user-images.githubusercontent.com/116039531/203897264-5711f884-8fc4-4ba2-8321-c99c4d87be29.png)

**Comment by awsproserve** [comment]
do you have something we can do with applovin that is similar?

**Comment by ritamsarmah** [comment]
Currently we only support event callbacks via C++ (e.g., binding to delegates like `UAppLovinMAX::OnInterstitialDisplayedDelegate`), though full blueprint support is something we are actively working on and plan to release very soon.

**Comment by ritamsarmah** [comment]
Hi @awsproserve, if you'd like to test out a version of the plugin with blueprint support, the feature is available under [this branch](https://github.com/AppLovin/AppLovin-MAX-Unreal/tree/feat/blueprint_delegates).

**Comment by awsproserve** [comment]
Wonderful, thank you so much @ritamsarmah I'll test this out and let you know. Is there release notes I can read to see what is new so I can compare, or is it the call back support for blueprint, just so I know what to test.

**Comment by ritamsarmah** [comment]
No release notes or docs yet, since that feature branch is still being finalized (and interface is subject to potential changes), but just wanted to give you the chance to try it out with your integration.

It includes a new `AppLovinMAXDelegate` blueprint class with dynamic delegates for each event. I'd recommend using delegates to determine when the SDK initializes before loading any ads and when ads finish loading before showing them. Here's a basic example for creating/showing a banner:

<img width="1148" alt="Screen Shot 2022-11-29 at 2 30 34 PM" src="https://user-images.githubusercontent.com/17148467/204662586-49749586-5208-4d72-8987-12ad50f6e57f.png">

**Comment by awsproserve** [comment]
Perfect, thank you, whats the best way to update the plugin, should I just delete the old one and reinstall?

**Comment by ritamsarmah** [comment]
Yep, that should work!

**Comment by NursultanNazarov** [comment]
> Yep, that should work!

Hi. I see you also used EasyAdsPro. I'm currently using it. Why did you choose Applovin MAX? Is it better?

**Comment by awsproserve** [comment]
> > Perfect, thank you, whats the best way to update the plugin, should I just delete the old one and reinstall?
> 
> Hi. I see you also used EasyAdsPro. I'm currently using it. Why did you choose Applovin MAX? Is it better?

Because EasyAdsPro you have to use mediation with googleadmob, and I have alot of issues with google admob, well, not only me but if you just google the problems you run into with them its alot and painful, and there is no support what so ever from google. I was using meta directly before they converted to solely working with mediation partners, meaning you can't use them stand alone, so we plan to use meta audience network along with Applovin. So far our experience with Applovin has been wonderful. We had 64 apps running on meta before with no problem (native IOS and Android apps), when meta changed to only work with mediation parters we went to google admob and thats where we ran into alot of issues, so we are going to migrate all our apps to use applovin. Not many mediation partners support unreal engine as well, so this is great that they are compatible with unreal engine! Hope that answers your question, let me know if you have any others, would be happy to share my experience, as we have used alot of different services these past few years.

**Comment by awsproserve** [comment]
> Yep, that should work!

wonderful trying this today, sorry been out sick, but im back today recovering slowly, will update you with my findings/experience.

**Comment by awsproserve** [comment]
> Yep, that should work!

@ritamsarmah can you send me the link again/ give me access, when I try the previous link it give me error on that page. Thank you.

**Comment by ritamsarmah** [comment]
Oops, the branch got deleted since we just merged it in. You can test the plugin straight off the `master` branch.

Also as a disclaimer, the plugin on `master` still isn't an "official" versioned release, so we may make interface/name changes in the meantime until the next major version update (aiming for the beginning of next year). However, the functionality you need is all fully implemented. :smile:

**Comment by awsproserve** [comment]
thanks for the heads up! diving into this right now! Many thanks for all your help and support!!!! A++++

**Comment by awsproserve** [comment]
@ritamsarmah Hi, I can't seem to get the function called "Add AppLovin Max Delegate" how to I find this function, see screenshot. 

![image](https://user-images.githubusercontent.com/116039531/205170577-092ea059-1482-48b8-a32d-cb239815dd1c.png)

**Comment by awsproserve** [comment]
I was able to setup everything else from your screenshot, except for the missing function for "Add AppLovin Max Delegate" 
![image](https://user-images.githubusercontent.com/116039531/205172178-b4f24a1d-e3e7-42df-8aa3-45b9fd7bc911.png)

**Comment by awsproserve** [comment]
I also am checking the demo app and also can't find the "Add AppLovin Max Delegate" function.

<img width="1486" alt="image" src="https://user-images.githubusercontent.com/116039531/205174990-d8ef40d8-4388-434c-bba7-bb3d9ced5531.png">

**Comment by ritamsarmah** [comment]
Hm, so currently `AppLovin MAX Delegate` is a blueprint component that needs to be instantiated and attached to a blueprint class, for example of an Actor.

<img width="1221" alt="Screen Shot 2022-12-01 at 2 38 41 PM" src="https://user-images.githubusercontent.com/17148467/205178591-59b82a4c-3f6f-4aa9-9b50-ed3316e378a7.png">

**Comment by awsproserve** [comment]
Hi @ritamsarmah Thanks for the response, im not sure if I am doing this correctly, please let me know. so this is what im doing.

Step 1: Create Blueprint, and choose actor component
Step 2: I've named my blueprint AppLovinMaxDelegate
![image](https://user-images.githubusercontent.com/116039531/205345497-3315db53-4fca-4089-98dd-1027bfcee12a.png)
Step3: open the blueprint you just created, and create a function
![image](https://user-images.githubusercontent.com/116039531/205345781-55ca8942-6ab2-4120-b3b3-419dc348e006.png)

Step 4: from he function type in Add Applov.... 
![image](https://user-images.githubusercontent.com/116039531/205346503-c36e24da-f027-43f2-b229-88d8e71faf9f.png)


Result, I am not able to see the AppLovin Max Delegate from the function im creating, as you have in your screenshot, Am I doing this right, 


Many thanks in advance you've great such amazing help so far!!! :)

**Comment by ritamsarmah** [comment]
I'll admit I'm not too well-versed in blueprints, but I think my approach would be to have a new or existing `Actor` blueprint class that handles the AppLovin MAX related code, and add the `AppLovin MAX Delegate` (which is already an actor component) to it.

**Comment by awsproserve** [comment]
Hi, sorry im alittle confused. im trying to see how to create the "AppLovin MAX Delegate", and what code AppLovin Max related code that you refer to above should go inside the "AppLovin MAX Delegate" function that im creating?

**Comment by awsproserve** [comment]
to clarify my question, see whats highlighted in teal blue boxes, that is the applovin max code, 
![image](https://user-images.githubusercontent.com/116039531/205512529-2643c1ea-d191-4def-8513-ad6e08ddcf7c.png)
so the question remains, what goes inside the AppLovin Max Delegate actor or function that im creating?

**Comment by awsproserve** [comment]
maybe you can send me your test file, and I can reverse engineer it from there to get the answer if that helps?

**Comment by awsproserve** [comment]
while packaging this latest plugin with the old configurations for initializing, banner, and interstitial, (the original instructions you gave me to create these, not the new ones because the new ones im missing the applovin max delegate actor/ function, but with the old configurations im getting red errors while packing the iOS package, starting with failed to initialize the plugin. 

here are a few of the errors we have, not all of them but some of them, but I do believe they start with the first error below.

im guessing the old setup doesn't work no more and I need to conform with the new setup/configurations. 

<img width="1350" alt="image" src="https://user-images.githubusercontent.com/116039531/205516708-76dafb58-6bae-4d82-aeb4-9868f065aae0.png">


here are some other errors,

<img width="1261" alt="image" src="https://user-images.githubusercontent.com/116039531/205516731-21407680-c099-4935-a305-7af521aaebf2.png">

<img width="1537" alt="image" src="https://user-images.githubusercontent.com/116039531/205516742-f70fb1b8-eca3-4339-a33f-5cd93e74965a.png">

<img width="1565" alt="image" src="https://user-images.githubusercontent.com/116039531/205516749-290467d9-9409-4815-8b37-35eac9275573.png">

<img width="1421" alt="image" src="https://user-images.githubusercontent.com/116039531/205516760-ae47ca52-91a7-455b-a74d-f8405cd8ea1a.png">


<img width="1408" alt="image" src="https://user-images.githubusercontent.com/116039531/205516775-b81140c9-9134-4971-921b-aec3a44f710a.png">


<img width="1546" alt="image" src="https://user-images.githubusercontent.com/116039531/205516790-733b038f-8bcf-4839-840e-4d38dcf1d39c.png">

**Comment by ritamsarmah** [comment]
Hi @awsproserve, you may try clearing all your cached build/intermediate directories before re-compiling and see if that gets rid of the errors.

I also apologize for the confusion about the integration. I'm taking some time to create a better example and write out some clearer instructions, and will report back once it's ready.

**Comment by awsproserve** [comment]
HI @ritamsarmah thanks so much that would be very much appreciated. We'll wait on those instructions. we cleaned the intermediate directories and we still come across the same errors, im assuming the new plugin has another method to initialize the plugin which is causing the errors as we are currenlty using the original instructions/ configurations.

**Comment by awsproserve** [comment]
> > > > Perfect, thank you, whats the best way to update the plugin, should I just delete the old one and reinstall?
> > > 
> > > 
> > > Hi. I see you also used EasyAdsPro. I'm currently using it. Why did you choose Applovin MAX? Is it better?
> > 
> > 
> > Because EasyAdsPro you have to use mediation with googleadmob, and I have alot of issues with google admob, well, not only me but if you just google the problems you run into with them its alot and painful, and there is no support what so ever from google. I was using meta directly before they converted to solely working with mediation partners, meaning you can't use them stand alone, so we plan to use meta audience network along with Applovin. So far our experience with Applovin has been wonderful. We had 64 apps running on meta before with no problem (native IOS and Android apps), when meta changed to only work with mediation parters we went to google admob and thats where we ran into alot of issues, so we are going to migrate all our apps to use applovin. Not many mediation partners support unreal engine as well, so this is great that they are compatible with unreal engine! Hope that answers your question, let me know if you have any others, would be happy to share my experience, as we have used alot of different services these past few years.
> 
> Thanks for reply! I would like to know more about applovin and how to setup mediation there. Currently my game reached 2000+ downloads a day. The only reason I'm considering Applovin is Yandex Ads and Admob Bidding(it's performing very good, like 10-14$ eCPM even though my top countries aren't Tier1) support. My game's 5% audience is russians. Can we talk in discord? My username in Dicord MadaraAlmighty#3477

just sent you a friend request via discord muzzy#6387

**Comment by ritamsarmah** [comment]
Here are the instructions I had for creating a new actor to implement the AppLovin MAX Delegate. I've also included the final blueprint for reference here: [Bp_AppLovinMAXDelegate.uasset.zip](https://github.com/AppLovin/AppLovin-MAX-Unreal/files/10160482/Bp_AppLovinMAXDelegate.uasset.zip)

1. Create a new `Actor` blueprint class that will implement the event callbacks.
2. Under the "Components" tab, click "Add Component" and search for "App Lovin MAX Delegate". I also made it the root component by dragging it up in that component hierarchy.
3. Right-click on the new component, go to "Add Event", and select the event to implement, e.g., "Add OnSdkInitializedDynamicDelegate". You can also add the event from the "Details" tab.
4. Implement the necessary logic branching off the newly created event node. Repeat for other ad events as needed.
5. This actor can then be spawned in your desired blueprint class. (Make sure you're also initializing the SDK somewhere)

<img width="1840" alt="Screen Shot 2022-12-05 at 5 58 32 PM" src="https://user-images.githubusercontent.com/17148467/205791969-c9c66776-f7a4-49e3-8fef-a8a2bd71b16c.png">

<img width="1840" alt="Screen Shot 2022-12-05 at 6 26 45 PM" src="https://user-images.githubusercontent.com/17148467/205793695-9f3eca61-5ba8-4121-a624-675454ef407e.png">

In general, you'll want to listen for the `OnSdkInitialized` event before loading an ad (whether it is banner, interstitial, etc.), and listen for the corresponding `On*AdLoaded` event before attempting to show it. You should also listen to the `On*AdFailedToDisplay` to implement any retry logic/error handling as desired.

I believe you could try adding the AppLovin MAX Delegate component as just a variable in other blueprint classes and implement the events in a similar way, but this was the example I had time to get working. If you need further clarification or have other suggestions for better integration steps, feel free to let me know.

**Comment by awsproserve** [comment]
Many thanks for the instructions, question how do you get the little computer screen on the Event On initialized in the second screenshot, and what is its purpose?
![image](https://user-images.githubusercontent.com/116039531/206064763-081d7f3b-b418-4f9a-932e-f589f4312250.png)

**Comment by awsproserve** [comment]
After setting it up as you described above, with the exception of the little computer screen icon that im missing as stated above when packaging I still get alot of red line errors related to the plugin, stating its not initialized properly
starting with this error:

![image](https://user-images.githubusercontent.com/116039531/206066773-f8e26374-4aad-40ec-9afc-77a210eef682.png)

**Comment by awsproserve** [comment]
I have setup in my global instance a custom event to initialize when called as in screenshot below 
![image](https://user-images.githubusercontent.com/116039531/206066947-92d732d4-cc0d-43ac-a9ae-d10a19addf84.png)

and on my first level blueprint in my begin play event, I call the custom event "custom event applovin initialize"

the custom event for initializing is connected to the event begin play in my first level blueprint 
![image](https://user-images.githubusercontent.com/116039531/206067132-8347f433-9a76-44fe-bd70-0719d0760004.png)

**Comment by ritamsarmah** [comment]
Hi @awsproserve, I've created a new branch with the fix for the errors you were encountering here: https://github.com/AppLovin/AppLovin-MAX-Unreal/tree/fix/default_ustruct_values

Would you mind trying that out (by copying the plugin from that branch) and letting me know if that fixes the issue?

**Comment by ritamsarmah** [comment]
Hmm yeah that's a really weird one. I'm looking into it right now, but one possible cause might be if bitcode is enabled in your project (since the newer builds we've been working on in master no longer have bitcode enabled in the iOS plugin). 

Could you try packaging after disabling that under Project Settings > Platforms - iOS > Build > Uncheck "Support bitcode in Shipping"? If that doesn't work, maybe you can try experimenting with the other options under the iOS build settings in the meantime while I also continue debugging it. I appreciate the patience and cooperation as we work out these issues—it's been equally helpful for us in order to improve the plugin!

![image](https://user-images.githubusercontent.com/17148467/206561267-b6b02368-d36d-4515-bab3-b12725666faf.png)

**Comment by awsproserve** [comment]
Hello thanks for the prompt response, I've checked our setup and it looks like we do have this unchecked already, we'll continue playing around with settings to see if we can get this working, let me know when you want us to try another build of the plugin and we will get on it asap, thanks so much in your partnership.

<img width="1501" alt="image" src="https://user-images.githubusercontent.com/116039531/206593113-3e968b76-a553-4790-a1d9-eb262dcbf65f.png">

**Comment by ritamsarmah** [comment]
Hi @awsproserve, would you mind pulling and test the latest change in the `fix/default_ustruct_values` branch? I tried re-exporting the framework that Unreal is complaining about.

If that doesn't work, could you verify if you're seeing the issue occur in our demo app (from the fix branch) as well?

**Comment by awsproserve** [comment]
Hi @ritamsarmah I just updated using the latest fix/default_ustruct_values branch you updated around 40 mins ago. to my app and still the same errors. I've also done the same for the demo project and I am still getting the same errors. here are screenshots while I tried to package your demo app

![image](https://user-images.githubusercontent.com/116039531/206821213-6a60e26c-5668-489d-89f6-9693febfd650.png)


We have also just updated our app using the old plugin file that worked originally and we are able to package successfully and see banner and interstitials in the game. (using the old working plugin)

**Comment by ritamsarmah** [comment]
Ah I think I found the reason why this issue happens. This seems to be because I compiled the latest version of the plugin with Xcode 14—which introduced those `msgSend` selector stubs and apparently prevents linking to the static framework with versions of Xcode 13 or lower. (Also why I probably couldn't reproduce this on my machine).

Could you see if upgrading to Xcode 14 fixes the issue? In the meantime, I'll look deeper into this and see whether maintaining compatibility with older Xcode is possible while compiling with the latest Xcode.

**Comment by ritamsarmah** [comment]
Hi @awsproserve, I was able to reproduce the errors you were seeing by switching to Xcode 13. I've added a re-compiled version of the framework in the same fix branch that should be compatible with Xcode 13 if you'd like to give it a try.

**Comment by awsproserve** [comment]
my apologies for the late folllow up, I've been traveling and hadn't got to it until today, Everything works PERFECT! YAY Thank you. I implemented the banner example you gave me and it works. im now working on testing fail node for banner and implementing interstitial. will keep you updated on my further testing :)

**Comment by awsproserve** [comment]
Question, do we implement the interstitial ad code in the same AppLovinMaxDelegate code we have the banners in?


<img width="1624" alt="image" src="https://user-images.githubusercontent.com/116039531/208215451-d72fb0ba-39f4-4cb4-8bef-6e33ab2056c6.png">

**Comment by awsproserve** [comment]
also would this be the right code to implement a retry to create and load banner when the original create and load banner fails?

![image](https://user-images.githubusercontent.com/116039531/208216235-c6fb3d79-7f1a-4d32-a38c-998d291672bc.png)

**Comment by ritamsarmah** [comment]
> Question, do we implement the interstitial ad code in the same AppLovinMaxDelegate code we have the banners in?

You could, though it's really up to you and the needs of your game if you want a new actor blueprint for separating out your interstitial logic. Here's a screenshot example you may use as reference for loading/showing an interstitial (including retries on fail with exponential delay):

<img width="1840" alt="Screen Shot 2022-12-19 at 1 50 58 PM" src="https://user-images.githubusercontent.com/17148467/208531519-ec0fab03-b234-4c53-bd8c-7f2d932bcdc3.png">

> also would this be the right code to implement a retry to create and load banner when the original create and load banner fails?

Hmm, couple things we'd want to fix there:

1. Are you using the logic in the second screenshot as well as the `Bp_AppLovinMAXDelegate` code? You only need to have one or the other, since they're both ways to bind to the events and create the banner.
2. We also wouldn't want to have `Bind Event to...` the `On Sdk Initialized` and `On Banner Ad Loaded` events twice. The "Bind Event.." nodes connected with the white lines don't indicate chronological execution order; instead, the important part is the logic coming off the red "Event" box which says: "Here is the logic I want to be triggered _every time_ the corresponding event occurs" (again this is equivalent to the method used in `Bp_AppLovinMAXDelegate` with those separate red event nodes). The current implementation would result in the logic being executed twice redundantly every time the delegate events are triggered.
3. Banners already have auto-refresh, so if that it sufficient for your use case, then you don't have to worry about implementing a manual retry.

**Comment by awsproserve** [comment]
> > Question, do we implement the interstitial ad code in the same AppLovinMaxDelegate code we have the banners in?
> 
> You could, though it's really up to you and the needs of your game if you want a new actor blueprint for separating out your interstitial logic. Here's a screenshot example you may use as reference for loading/showing an interstitial (including retries on fail with exponential delay):
> 
> <img alt="Screen Shot 2022-12-19 at 1 50 58 PM" width="1840" src="https://user-images.githubusercontent.com/17148467/208531519-ec0fab03-b234-4c53-bd8c-7f2d932bcdc3.png">
> 

Thanks so much for your response, question are these custom events that you are creating for the interstitial, I tried creating one but it looks alittle different. different meaning my event says custom event on it, and yours does not. see screenshot below: 

purple box shows the custom event I created

![image](https://user-images.githubusercontent.com/116039531/208798993-abad9fe1-3a6c-49b0-a1fb-5454c86bdc5e.png)

**Comment by awsproserve** [comment]
the next screenshot here, shows me looking for the event you created on sdk initialized dynamic delegate. is this the way I should be setting these events up?

<img width="1267" alt="image" src="https://user-images.githubusercontent.com/116039531/208799330-40e5e678-8569-4cae-9954-60249baab498.png">

**Comment by awsproserve** [comment]
> Hmm, couple things we'd want to fix there:
> 
> 1. Are you using the logic in the second screenshot as well as the `Bp_AppLovinMAXDelegate` code? You only need to have one or the other, since they're both ways to bind to the events and create the banner.
> 2. We also wouldn't want to have `Bind Event to...` the `On Sdk Initialized` and `On Banner Ad Loaded` events twice. The "Bind Event.." nodes connected with the white lines don't indicate chronological execution order; instead, the important part is the logic coming off the red "Event" box which says: "Here is the logic I want to be triggered _every time_ the corresponding event occurs" (again this is equivalent to the method used in `Bp_AppLovinMAXDelegate` with those separate red event nodes). The current implementation would result in the logic being executed twice redundantly every time the delegate events are triggered.
> 3. Banners already have auto-refresh, so if that is sufficient for your use case, then you don't have to worry about implementing a manual retry.


as for my banner setup, I have a event begin play in my level blueprint "level blueprint called gangfembunny", that calls on an event I have in my global game instance to initialize the plugin.

![image](https://user-images.githubusercontent.com/116039531/208800776-3461093f-498e-4c49-935d-58b2658c471f.png)

here is the code in my global instance that initializes the plugin

![image](https://user-images.githubusercontent.com/116039531/208800977-fed16939-b6b9-4751-b0b0-959360632a6f.png)


after that I have it connected to the add applovin max delegate as seen in the screenshot below

![image](https://user-images.githubusercontent.com/116039531/208801106-9007c485-23bc-4dfa-b78b-0f77e8605559.png)

this is screenshot inside the applovin max delegate

![image](https://user-images.githubusercontent.com/116039531/208801246-e7212a6c-8a3d-4ce8-b666-fd8826022b03.png)


then I have all the binding stuff connected to the applovin max delegate

![image](https://user-images.githubusercontent.com/116039531/208801408-ee6646b0-d432-481f-ae76-902b7ba9aab9.png)


is this the correct setup, as this was the setup that you originally prescribed to me

**Comment by ritamsarmah** [comment]
> Thanks so much for your response, question are these custom events that you are creating for the interstitial, I tried creating one but it looks alittle different. different meaning my event says custom event on it, and yours does not. see screenshot below:

Not sure if there's a functional difference, but as long as you get it to work that's fine! If you're looking for the exact way I did it, I clicked on "AppLovinMAXDelegate" component in the left side and it shows all the events under "Details" on the right. You can click the green "+" button next to the event name to bind to it.

<img width="1840" alt="Screen Shot 2022-12-21 at 9 06 45 AM" src="https://user-images.githubusercontent.com/17148467/208963276-daa36d7b-2de5-4841-873e-e4498635f237.png">

> is this the correct setup, as this was the setup that you originally prescribed to me as seen in the screenshot of your recommendations below: I just want to make sure im understanding this correctly. many thanks in advance for all your help.

The setup you have in your level blueprint seems good to me; however, you won't even need the `Bp_AppLovinMAXDelegate` in that case since it's just different way to do the same thing. They're both implementing the same logic for creating a banner as you can see, so you only need to select one method that suits your needs. If you're setting up the "AppLovin MAX Delegate" directly in your level blueprint,

**Comment by awsproserve** [comment]
Thanks so much, this was very helpful, I was able to set it up just as you did. the way I was doing was not right with custom event method. 

One last questions.

so is this correct, that now that I have the interstitial delegate BP setup, so now all I have to do is spawn that delegate actor, and it will listen when I call show interstitial from another level?

So what im going to do is in my main menu, (called gangfembunny) I will call an event from my global instance as seen in the screenshot below to spawn the actor for the interstitial delegate 
![image](https://user-images.githubusercontent.com/116039531/209230482-d303e076-106b-439c-8f78-e1729ad27c32.png)


Then later on in another level I will call show interstitial. Is that the correct setup? or do I have to spawn the interstitial delegate in the same level that I'll be showing the interstitial ad?

**Comment by awsproserve** [comment]
also what delegates would I use in the screenshot below if:

User clicks the x during the interstitial ad and exits
Interstitial ad plays out and ends without user clicking the add. ie it naturally exits

The reason is because after the interstitial ad is done, I want to have it restart the level. So wanting to see which I should call on.

![image](https://user-images.githubusercontent.com/116039531/209230934-bd7cdaa4-7112-4970-b8ba-006340f4b231.png)

**Comment by awsproserve** [comment]
I just packaged app and tested it, it looks like when I implement this code in screenshot below, the interstitial ad looks like it only loads one time, and thats when I initially spawn the blueprint in my main menu, then the player goes into level 1 and dies and then calls to show the ad. and it shows. but then when the player dies again it doesn't show the ad any more, ie its not loading another ad after the initial show of the ad.


<img width="1394" alt="image" src="https://user-images.githubusercontent.com/116039531/209330402-1a478d3e-4b3f-4212-a49b-278b0ed59bc6.png">

**Comment by awsproserve** [comment]
I added the two delegates below in the box, so the interstitial can reload after being displayed or clicked and its still not reloading

![image](https://user-images.githubusercontent.com/116039531/209412187-2c544455-e3b1-4618-8d8f-171544b8c1f7.png)

**Comment by ritamsarmah** [comment]
Your approach sounds good so far, though you shouldn't be loading after "On Interstitial Clicked" and "On Interstitial Displayed"—adding the load after "On Interstitial Hidden" is sufficient.

I'd suggest checking the device logs or adding a log after the events, namely `On Interstitial Hidden` and `On Interstitial Loaded` to check if the callback is being invoked or not. Another thing to check is if the next interstitial is not getting loaded in time, by adding a log statement to the `False` condition after `Is Interstitial Ready`.

**Comment by ritamsarmah** [comment]
Hi @awsproserve, just following up on this recent interstitial loading issue to see if you were able to resolve it? If not, we can continue the discussion over email, but I plan on closing this ticket out in any case since the original issue has been addressed.

**Comment by awsproserve** [comment]
Hi, thanks for following up, Sure thing we can continue discussion over email. I was traveling for the holidays and playing catch up now, I’ll make sure to have this tested by end of week and will report back. Thanks again for everything.

**Comment by ritamsarmah** [comment]
Sounds great, feel free to email me at ritam.sarmah@applovin.com
